File: go/modcdp/transport/ReverseWebSocketUpstreamTransport.go
package transport

import (
	"encoding/json"
	"fmt"
	"net"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/gobwas/ws"
	"github.com/gobwas/ws/wsutil"
)

const DefaultUpstreamReverseWSBind = "127.0.0.1:29292"
const DefaultUpstreamReverseWSWaitTimeoutMS = 10_000

type ReverseWebSocketUpstreamTransport struct {
	UpstreamTransport
	Bind          string
	URL           string
	WaitTimeoutMS int
	Listener      net.Listener
	Conn          net.Conn
	writeMu       sync.Mutex
	peerCh        chan struct{}
	peerOnce      sync.Once
	closeCh       chan struct{}
	stateMu       sync.Mutex
	PeerInfo      map[string]any
	generation    int64
}

type ReverseWebSocketUpstreamTransportOptions struct {
	UpstreamReverseWSBind          string `json:"upstream_reversews_bind,omitempty"`
	UpstreamReverseWSWaitTimeoutMS int    `json:"upstream_reversews_wait_timeout_ms,omitempty"`
}

func NewReverseWebSocketUpstreamTransport(options ReverseWebSocketUpstreamTransportOptions) *ReverseWebSocketUpstreamTransport {
	reverseWSBind := options.UpstreamReverseWSBind
	if reverseWSBind == "" {
		reverseWSBind = DefaultUpstreamReverseWSBind
	}
	reverseWSWaitTimeoutMS := options.UpstreamReverseWSWaitTimeoutMS
	if reverseWSWaitTimeoutMS == 0 {
		reverseWSWaitTimeoutMS = DefaultUpstreamReverseWSWaitTimeoutMS
	}
	t := &ReverseWebSocketUpstreamTransport{WaitTimeoutMS: reverseWSWaitTimeoutMS, peerCh: make(chan struct{}), closeCh: make(chan struct{})}
	t.setBind(reverseWSBind)
	return t
}

func (t *ReverseWebSocketUpstreamTransport) setBind(bind string) {
	raw := bind
	if !strings.Contains(raw, "://") {
		raw = "ws://" + raw
	}
	parsed, err := url.Parse(raw)
	if err != nil {
		parsed, _ = url.Parse("ws://" + DefaultUpstreamReverseWSBind)
	}
	host := parsed.Hostname()
	if host == "" {
		host = "127.0.0.1"
	}
	port := parsed.Port()
	if port == "" {
		port = "29292"
	}
	t.Bind = net.JoinHostPort(host, port)
	t.URL = "ws://" + t.Bind
}

func (t *ReverseWebSocketUpstreamTransport) Update(config map[string]any) {
	if config == nil {
		return
	}
	if bind, _ := config["upstream_reversews_bind"].(string); bind != "" {
		t.setBind(bind)
	}
	if waitTimeoutMS, ok := intFromConfig(config["upstream_reversews_wait_timeout_ms"]); ok {
		t.WaitTimeoutMS = waitTimeoutMS
	}
}

func (t *ReverseWebSocketUpstreamTransport) Connect() error {
	t.stateMu.Lock()
	t.closeCh = make(chan struct{})
	t.stateMu.Unlock()
	listener, err := net.Listen("tcp", t.Bind)
	if err != nil {
		return err
	}
	t.Listener = listener
	go t.acceptLoop()
	return nil
}

func (t *ReverseWebSocketUpstreamTransport) Send(message map[string]any) error {
	body, err := json.Marshal(message)
	if err != nil {
		return err
	}
	t.writeMu.Lock()
	defer t.writeMu.Unlock()
	conn := t.Conn
	if conn == nil {
		return fmt.Errorf("no reverse ModCDP extension peer is connected at %s", t.URL)
	}
	return wsutil.WriteServerText(conn, body)
}

func (t *ReverseWebSocketUpstreamTransport) GetInjectorConfig() ExtensionInjectorConfig {
	return ExtensionInjectorConfig{}
}

func (t *ReverseWebSocketUpstreamTransport) WaitForPeer() error {
	t.writeMu.Lock()
	connected := t.Conn != nil
	t.writeMu.Unlock()
	if connected {
		return nil
	}
	t.stateMu.Lock()
	closeCh := t.closeCh
	peerCh := t.peerCh
	t.stateMu.Unlock()
	select {
	case <-peerCh:
		return nil
	case <-closeCh:
		return fmt.Errorf("reverse websocket transport at %s closed before a peer connected", t.URL)
	case <-time.After(time.Duration(t.WaitTimeoutMS) * time.Millisecond):
		return fmt.Errorf("timed out waiting %dms for reverse ModCDP extension connection", t.WaitTimeoutMS)
	}
}

func (t *ReverseWebSocketUpstreamTransport) PeerGeneration() int64 {
	t.stateMu.Lock()
	defer t.stateMu.Unlock()
	return t.generation
}

func (t *ReverseWebSocketUpstreamTransport) Close() error {
	t.stateMu.Lock()
	closeCh := t.closeCh
	t.closeCh = make(chan struct{})
	t.peerCh = make(chan struct{})
	t.peerOnce = sync.Once{}
	t.stateMu.Unlock()
	close(closeCh)
	t.writeMu.Lock()
	if t.Conn != nil {
		_ = t.Conn.Close()
		t.Conn = nil
	}
	t.writeMu.Unlock()
	if t.Listener != nil {
		_ = t.Listener.Close()
		t.Listener = nil
	}
	t.stateMu.Lock()
	t.PeerInfo = nil
	t.stateMu.Unlock()
	return nil
}

func (t *ReverseWebSocketUpstreamTransport) acceptLoop() {
	for {
		conn, err := t.Listener.Accept()
		if err != nil {
			return
		}
		go t.accept(conn)
	}
}

func (t *ReverseWebSocketUpstreamTransport) accept(conn net.Conn) {
	if _, err := ws.Upgrade(conn); err != nil {
		_ = conn.Close()
		t.emitClose(err)
		return
	}
	_ = conn.SetReadDeadline(time.Now().Add(time.Duration(t.WaitTimeoutMS) * time.Millisecond))
	helloBytes, err := wsutil.ReadClientText(conn)
	_ = conn.SetReadDeadline(time.Time{})
	if err != nil {
		_ = conn.Close()
		t.emitClose(err)
		return
	}
	var hello map[string]any
	if err := json.Unmarshal(helloBytes, &hello); err != nil || hello["type"] != "modcdp.reverse.hello" {
		_ = conn.Close()
		if err == nil {
			err = fmt.Errorf("invalid reverse hello")
		}
		t.emitClose(err)
		return
	}
	t.writeMu.Lock()
	if t.Conn != nil && t.Conn != conn {
		_ = t.Conn.Close()
	}
	t.Conn = conn
	t.writeMu.Unlock()
	t.stateMu.Lock()
	t.PeerInfo = hello
	t.generation++
	t.peerOnce.Do(func() { close(t.peerCh) })
	t.stateMu.Unlock()
	for {
		data, err := wsutil.ReadClientText(conn)
		if err != nil {
			t.writeMu.Lock()
			if t.Conn == conn {
				t.Conn = nil
				t.writeMu.Unlock()
				t.stateMu.Lock()
				t.PeerInfo = nil
				t.stateMu.Unlock()
				t.resetPeerWait()
			} else {
				t.writeMu.Unlock()
			}
			t.emitClose(err)
			return
		}
		var message map[string]any
		if err := json.Unmarshal(data, &message); err == nil {
			t.emitRecv(message)
		}
	}
}

func (t *ReverseWebSocketUpstreamTransport) resetPeerWait() {
	t.stateMu.Lock()
	defer t.stateMu.Unlock()
	t.peerCh = make(chan struct{})
	t.peerOnce = sync.Once{}
}

File: python/modcdp/transport/ReverseWebSocketUpstreamTransport.py
from __future__ import annotations

import base64
import hashlib
import json
import socket
import struct
import threading
import time
from typing import Any
from urllib.parse import urlparse

from ..transport.UpstreamTransport import UpstreamTransport


DEFAULT_UPSTREAM_REVERSEWS_BIND = "127.0.0.1:29292"
DEFAULT_UPSTREAM_REVERSEWS_WAIT_TIMEOUT_MS = 10_000
_WS_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"


class ReverseWebSocketUpstreamTransport(UpstreamTransport):
    mode = "reversews"
    endpoint_kind = "modcdp_server"

    def __init__(self, options: dict[str, Any] | None = None) -> None:
        super().__init__()
        options = options or {}
        self.wait_timeout_ms = int(options.get("upstream_reversews_wait_timeout_ms") or DEFAULT_UPSTREAM_REVERSEWS_WAIT_TIMEOUT_MS)
        self.server_socket: socket.socket | None = None
        self.socket: socket.socket | None = None
        self.accept_thread: threading.Thread | None = None
        self.reader_thread: threading.Thread | None = None
        self.peer_info: dict[str, Any] | None = None
        self.peer_event = threading.Event()
        self._peer_condition = threading.Condition()
        self._close_generation = 0
        self.closed = False
        self.write_lock = threading.Lock()
        self._setBind(str(options.get("upstream_reversews_bind") or DEFAULT_UPSTREAM_REVERSEWS_BIND))

    def update(self, config: dict[str, Any] | None = None) -> "ReverseWebSocketUpstreamTransport":
        config = config or {}
        bind = config.get("upstream_reversews_bind")
        if isinstance(bind, str) and bind:
            self._setBind(bind)
        wait_timeout_ms = config.get("upstream_reversews_wait_timeout_ms")
        if isinstance(wait_timeout_ms, int | float):
            self.wait_timeout_ms = int(wait_timeout_ms)
        return self

    def getInjectorConfig(self) -> dict[str, Any]:
        return {}

    def _setBind(self, bind: str) -> None:
        parsed = urlparse(bind if "://" in bind else f"ws://{bind}")
        host = parsed.hostname or "127.0.0.1"
        port = parsed.port or 29292
        if port <= 0 or port > 65535:
            raise ValueError(f"Invalid host:port {bind}")
        self.url = f"ws://{host}:{port}"

    def connect(self) -> None:
        parsed = urlparse(self.url or "")
        host = parsed.hostname or "127.0.0.1"
        port = parsed.port or 29292
        server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind((host, port))
        server_socket.listen(1)
        self.server_socket = server_socket
        with self._peer_condition:
            self._close_generation += 1
            self.peer_event.clear()
        self.closed = False
        self.accept_thread = threading.Thread(target=self._accept_loop, daemon=True)
        self.accept_thread.start()

    def send(self, message: dict[str, Any]) -> None:
        if self.socket is None:
            raise RuntimeError(f"No reverse ModCDP extension peer is connected at {self.url}.")
        payload = json.dumps(message).encode()
        with self.write_lock:
            self.socket.sendall(_encode_server_text_frame(payload))

    def waitForPeer(self) -> None:
        if self.socket is not None:
            return
        deadline = time.monotonic() + self.wait_timeout_ms / 1000
        with self._peer_condition:
            close_generation = self._close_generation
            while self.socket is None:
                if close_generation != self._close_generation:
                    raise RuntimeError(f"Reverse websocket transport at {self.url} closed before a peer connected.")
                remaining = deadline - time.monotonic()
                if remaining <= 0:
                    raise RuntimeError(f"Timed out waiting {self.wait_timeout_ms}ms for reverse ModCDP extension connection.")
                self._peer_condition.wait(remaining)

    def close(self) -> None:
        self.closed = True
        for sock in (self.socket, self.server_socket):
            try:
                sock.close() if sock is not None else None
            except Exception:
                pass
        self.socket = None
        self.server_socket = None
        self.peer_info = None
        self.peer_event.clear()
        with self._peer_condition:
            self._close_generation += 1
            self._peer_condition.notify_all()

    def _accept_loop(self) -> None:
        while not self.closed and self.server_socket is not None:
            try:
                sock, _ = self.server_socket.accept()
                self._accept(sock)
            except OSError:
                return
            except Exception as error:
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _accept(self, sock: socket.socket) -> None:
        try:
            _perform_server_handshake(sock)
            hello_raw = _read_client_text_frame(sock, self.wait_timeout_ms / 1000)
            if hello_raw is None:
                raise RuntimeError("reverse hello socket closed")
            hello = json.loads(hello_raw)
            if not isinstance(hello, dict) or hello.get("type") != "modcdp.reverse.hello":
                raise RuntimeError("invalid reverse hello")
            old_socket = self.socket
            if old_socket is not None and old_socket is not sock:
                try:
                    old_socket.close()
                except Exception:
                    pass
            with self._peer_condition:
                self.socket = sock
                self.peer_info = hello
                self.peer_event.set()
                self._peer_condition.notify_all()
            self.reader_thread = threading.Thread(target=self._read_loop, args=(sock,), daemon=True)
            self.reader_thread.start()
        except Exception as error:
            try:
                sock.close()
            except Exception:
                pass
            self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))

    def _read_loop(self, sock: socket.socket) -> None:
        try:
            while not self.closed and self.socket is sock:
                data = _read_client_text_frame(sock, None)
                if data is None:
                    break
                self._parse_and_emit_recv(data)
        except Exception as error:
            if not self.closed:
                self._emit_close(error if isinstance(error, Exception) else RuntimeError(str(error)))
        finally:
            if self.socket is sock:
                with self._peer_condition:
                    if self.socket is sock:
                        self.socket = None
                        self.peer_info = None
                        self.peer_event.clear()
                        self._peer_condition.notify_all()
            try:
                sock.close()
            except Exception:
                pass


def _perform_server_handshake(sock: socket.socket) -> None:
    request = b""
    while b"\r\n\r\n" not in request:
        chunk = sock.recv(4096)
        if not chunk:
            raise RuntimeError("websocket handshake closed")
        request += chunk
    headers: dict[str, str] = {}
    for line in request.decode(errors="replace").split("\r\n")[1:]:
        if ":" in line:
            key, value = line.split(":", 1)
            headers[key.strip().lower()] = value.strip()
    ws_key = headers.get("sec-websocket-key")
    if not ws_key:
        raise RuntimeError("websocket handshake missing Sec-WebSocket-Key")
    accept = base64.b64encode(hashlib.sha1((ws_key + _WS_GUID).encode()).digest()).decode()
    sock.sendall(
        (
            "HTTP/1.1 101 Switching Protocols\r\n"
            "Upgrade: websocket\r\n"
            "Connection: Upgrade\r\n"
            f"Sec-WebSocket-Accept: {accept}\r\n"
            "\r\n"
        ).encode()
    )


def _read_exact(sock: socket.socket, length: int) -> bytes:
    chunks: list[bytes] = []
    remaining = length
    while remaining > 0:
        chunk = sock.recv(remaining)
        if not chunk:
            raise EOFError("websocket closed")
        chunks.append(chunk)
        remaining -= len(chunk)
    return b"".join(chunks)


def _read_client_text_frame(sock: socket.socket, timeout_s: float | None) -> str | None:
    old_timeout = sock.gettimeout()
    if timeout_s is not None:
        sock.settimeout(timeout_s)
    try:
        header = _read_exact(sock, 2)
        opcode = header[0] & 0x0F
        masked = bool(header[1] & 0x80)
        length = header[1] & 0x7F
        if length == 126:
            length = struct.unpack("!H", _read_exact(sock, 2))[0]
        elif length == 127:
            length = struct.unpack("!Q", _read_exact(sock, 8))[0]
        mask = _read_exact(sock, 4) if masked else b""
        payload = bytearray(_read_exact(sock, length))
        if masked:
            for index, value in enumerate(payload):
                payload[index] = value ^ mask[index % 4]
        if opcode == 0x8:
            return None
        if opcode != 0x1:
            return ""
        return payload.decode()
    finally:
        sock.settimeout(old_timeout)


def _encode_server_text_frame(payload: bytes) -> bytes:
    if len(payload) < 126:
        header = bytes([0x81, len(payload)])
    elif len(payload) < 65536:
        header = bytes([0x81, 126]) + struct.pack("!H", len(payload))
    else:
        header = bytes([0x81, 127]) + struct.pack("!Q", len(payload))
    return header + payload
